course6/string_ariel/str.h

```diff
@@ -2,192 +2,100 @@
         {
-            *end++ = str[i];
+            Initialize(s);
         }
 
-        mid = end;
-    }
-
-    Str(Str&& str) :start(str.start), mid(str.mid), end(str.end)
-    {
-        str.start = str.mid = str.end = nullptr;
-    }
-
-    ~Str()
-    {
-        Dispose();
-    }
-
-    char operator[](int i) const
-    {
-        return *(start + i);
-    }
-
-    char& operator[](int i)
-    {
-        return *(start + i);
-    }
-
-    Str& operator = (const Str& str)
-    {
-        if (str.Size() > Capacity())
+        Str(const Str& str)
         {
-            Str temp(str);
-            Swap(*this, temp);
+            Initialize(str.start);
         }
-        else
-        {
-            mid = start + str.Size();
-            Copy(str);
-        }
-
-        return *this;
-    }
 
-    Str& operator = (const char* s)
-    {
-        int size = std::strlen(s);
-        if (size > Capacity())
+        Str(Str&& str) : start(str.start), mid(str.mid), end(str.end)
         {
-            Str temp(s);
-            Swap(*this, temp);
+            str.start = str.mid = str.end = nullptr;
         }
-        else
+
+        ~Str()
         {
-            mid = start + size;
-            Copy(s);
+            delete[] start;
         }
 
-        return *this;
-    }
-
-    Str operator + (const Str& rh)
-    {
-        return Str(*this, rh);
-    }
-
-    Str& operator += (const Str& rh)
-    {
-        if(rh.Size() + Size() <= Capacity())
+        char operator[](int i) const
         {
-            for(int i = 0; i < rh.Size(); i++)
-            {
-                *mid++ = rh[i];
-            }
+            return start[i];
         }
-        else
+
+        char& operator[](int i)
         {
-            Str temp = *this + rh;
-            Swap(*this, temp);
+            return start[i];
         }
 
-        return *this;
-    }
-
-    int Size() const
-    {
-        return mid - start;
-    }
-
-    int Capacity() const
-    {
-        return end - start;
-    }
-
-    friend std::ostream &operator<<(std::ostream &os, const Str& str);
-
-private:
-    char* start, *end, *mid;
-
-    Str(const Str& str1, const Str& str2)
-    {
-        int size1 = str1.Size();
-        int size2 = str2.Size();
+        Str& operator = (const Str& str)
+        {
+            Assign(str.start, str.size());
 
-        start = new char[size1 + size2];
-        end = start;
+            return *this;
+        }
 
-        for (int i = 0; i < size1; i++)
+        Str& operator = (const char* s)
         {
-            *end++ = str1[i];
+            Assign(s, std::strlen(s));
+
+            return *this;
         }
 
-        for (int i = 0; i < size2; i++)
+        Str operator + (const Str& rh)
         {
-            *end++ = str2[i];
+            return Str(*this, rh);
         }
 
-        mid = end;
-    }
+        Str& operator += (const Str& rh)
+        {
+            Assign((*this + rh).start, this->size() + rh.size());
```

Comment:
I know this is an extra allocatoin/deallocation but makes the code much more readable